test/queues.c

```diff
@@ -166,8 +166,116 @@ TEST test_queue_teardown(void)
+
+        struct notification *n = test_notification("n", 0);
+
+        queues_notification_insert(n);
+        queues_update(STATUS_NORMAL);
+
+        ASSERTm("Age threshold is deactivated and the notification is infinite, there is no wakeup necessary.",
+                queues_get_next_datachange(time_monotonic_now()) < 0);
+
+        queues_teardown();
+        PASS();
+}
+
+TEST test_datachange_endless_agethreshold(void)
+{
+        settings.show_age_threshold = S2US(5);
+
+        queues_init();
+
+        struct notification *n = test_notification("n", 0);
+
+        queues_notification_insert(n);
+        queues_update(STATUS_NORMAL);
+
+        ASSERT_IN_RANGEm("Age threshold is activated and the next wakeup should be less than a second away",
+                S2US(1), queues_get_next_datachange(time_monotonic_now() + S2US(4)), S2US(1));
+
+        ASSERT_IN_RANGEm("Age threshold is activated and the next wakeup should be less than the age threshold",
+                settings.show_age_threshold, queues_get_next_datachange(time_monotonic_now()), settings.show_age_threshold);
+
+        settings.show_age_threshold = S2US(0);
+        ASSERT_IN_RANGEm("Age threshold is activated and the next wakeup should be less than a second away",
+                S2US(1), queues_get_next_datachange(time_monotonic_now()), S2US(1));
+
+        queues_teardown();
+        PASS();
+}
+
+TEST test_datachange_queues(void)
+{
+        queues_init();
+
+        struct notification *n = test_notification("n", 10);
+
+        queues_notification_insert(n);
+        ASSERTm("The inserted notification is inside the waiting queue, so it should get ignored.",
+               queues_get_next_datachange(time_monotonic_now()) < S2US(0));
+
+        queues_update(STATUS_NORMAL);
+        ASSERT_IN_RANGEm("The notification has to get closed in less than its timeout",
+               S2US(10), queues_get_next_datachange(time_monotonic_now()), S2US(10));
+
+        queues_notification_close(n, REASON_UNDEF);
+        ASSERTm("The inserted notification is inside the history queue, so it should get ignored",
+               queues_get_next_datachange(time_monotonic_now()) < S2US(0));
+
+        queues_teardown();
+        PASS();
+}
+
+TEST test_datachange_ttl(void)
+{
+        struct notification *n;
+        queues_init();
+
+        n = test_notification("n1", 15);
+
+        queues_notification_insert(n);
+        queues_update(STATUS_NORMAL);
+        ASSERT_IN_RANGEm("The notification has to get closed in less than its timeout.",
+               S2US(n->timeout), queues_get_next_datachange(time_monotonic_now()), S2US(n->timeout));
+
+        n = test_notification("n2", 10);
+
+        queues_notification_insert(n);
+        queues_update(STATUS_NORMAL);
+        ASSERT_IN_RANGEm("The timeout of the second notification has to get used as sleep time now.",
+               S2US(n->timeout), queues_get_next_datachange(time_monotonic_now()), S2US(n->timeout));
```

Comment:
Same thing in this test.

Reply:
There's something worse. Do you spot the problem?

Reply:
Oh, calling `S2US` on `timeout` nice catch!